### PR TITLE
Add table aws_glue_catalog_database. closes #201

### LINKED ISSUE #201: Add table aws_glue_database

### PR DESCRIPTION
**Pull request by pskrbasu**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT 300

SETUP: tests/aws_glue_catalog_database []

PRETEST: tests/aws_glue_catalog_database

TEST: tests/aws_glue_catalog_database
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_partition.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_glue_catalog_database.named_test_resource: Creating...
aws_glue_catalog_database.named_test_resource: Creation complete after 2s [id=123456789012:turbottest33108]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

aws_account = 123456789012
aws_partition = aws
aws_region = us-east-1
resource_aka = arn:aws:glue:us-east-1:123456789012:database/turbottest33108
resource_name = turbottest33108

Running SQL query: test-get-query.sql
[
  {
    "akas": [
      "arn:aws:glue:us-east-1:123456789012:database/turbottest33108"
    ],
    "catalog_id": "123456789012",
    "create_table_default_permissions": [
      {
        "Permissions": [
          "ALL"
        ],
        "Principal": {
          "DataLakePrincipalIdentifier": "IAM_ALLOWED_PRINCIPALS"
        }
      }
    ],
    "description": "integration testing",
    "location_uri": null,
    "name": "turbottest33108",
    "title": "turbottest33108"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "arn:aws:glue:us-east-1:123456789012:database/turbottest33108"
    ],
    "title": "turbottest33108"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "description": "integration testing",
    "name": "turbottest33108",
    "partition": "aws",
    "region": "us-east-1",
    "title": "turbottest33108"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "arn:aws:glue:us-east-1:123456789012:database/turbottest33108"
    ],
    "name": "turbottest33108",
    "region": "us-east-1",
    "title": "turbottest33108"
  }
]
✔ PASSED

POSTTEST: tests/aws_glue_catalog_database

TEARDOWN: tests/aws_glue_catalog_database

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
> select
  name,
  catalog_id,
  create_time,
  description,
  location_uri,
  create_table_default_permissions
from
  aws_glue_catalog_database;
+----------+--------------+---------------------+--------------------------+----------------+--------------------------------------------------------------------------
| name     | catalog_id   | create_time         | description              | location_uri   | create_table_default_permissions                                         
+----------+--------------+---------------------+--------------------------+----------------+--------------------------------------------------------------------------
| test_db2 | 453319552164 | 2021-04-27 12:43:30 | testing steampipe tables | www.turbot.com | [{"Permissions":["ALL"],"Principal":{"DataLakePrincipalIdentifier":"IAM_A
| test_db  | 453319552164 | 2021-04-27 06:43:31 | test                     | <null>         | [{"Permissions":["ALL"],"Principal":{"DataLakePrincipalIdentifier":"IAM_A
+----------+--------------+---------------------+--------------------------+----------------+--------------------------------------------------------------------------
> select
  catalog_id,
  count(name)
from
  aws_glue_catalog_database
group by
  catalog_id;
+--------------+-------+
| catalog_id   | count |
+--------------+-------+
| 453319552164 | 2     |
+--------------+-------+
> 

</details>

### CODE REVIEW

File: aws-test/tests/aws_glue_catalog_database/variables.tf [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+variable "resource_name" {
+  type    = string
+  default = "turbot-test-20200125-create-update"
+  description = "Name of the resource used throughout the test."
+}
+
+variable "aws_profile" {
+  type    = string
+  default = "integration-tests"
```
Comment (khushboo9024) [inline aws-test/tests/aws_glue_catalog_database/variables.tf:11]:
```suggestion
  default = "default"
```

File: aws/plugin.go [hunk 2]
```diff
@@ -86,6 +86,7 @@ func Plugin(ctx context.Context) *plugin.Plugin {
 			"aws_emr_cluster":                        tableAwsEmrCluster(ctx),
 			"aws_eventbridge_rule":                   tableAwsEventBridgeRule(ctx),
 			"aws_glacier_vault":                      tableAwsGlacierVault(ctx),
+			"aws_glue_catalog_database":							tableAwsGlueDatabase(ctx),
```
Comment (khushboo9024) [inline aws/plugin.go:89]: indentation is wrong.

File: aws/table_aws_glue_catalog_database.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+				Description: "A description of the database.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "location_uri",
+				Description: "The location of the database (for example, an HDFS path).",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "name",
```
Comment (khushboo9024) [inline aws/table_aws_glue_catalog_database.go:54]: `name` should be the first column.

File: aws/table_aws_glue_catalog_database.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+			{
+				Name:        "parameters",
+				Description: "These key-value pairs define parameters and properties of the database.",
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "target_database",
+				Description: "A DatabaseIdentifier structure that describes a target database for resource linking.",
+				Type:        proto.ColumnType_JSON,
+			},
```
Comment (khushboo9024) [inline aws/table_aws_glue_catalog_database.go:67]:
```suggestion
			},
			
```

File: aws/table_aws_glue_catalog_database.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+		},
+		GetMatrixItem: BuildRegionList,
+		Columns: awsRegionalColumns([]*plugin.Column{
+			{
+				Name:        "catalog_id",
+				Description: "The ID of the Data Catalog in which the database resides.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "create_table_default_permissions",
```
Comment (khushboo9024) [inline aws/table_aws_glue_catalog_database.go:34]: Ordering of column in not correct.

File: aws/table_aws_glue_catalog_database.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	if strings.HasPrefix(name, "/") {
+		aka = aka + name
+	} else {
+		aka = aka + "/" + name
+	}
+
+	return []string{aka}, nil
+}
+
+func databaseName(item interface{}) string {
```
Comment (khushboo9024) [inline aws/table_aws_glue_catalog_database.go:177]: This function is not required, we have same output for list and get call.

File: aws/table_aws_glue_catalog_database.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+				Transform:   transform.FromField("Name"),
+			},
+		}),
+	}
+}
+
+//// LIST FUNCTION
+
+func listAwsGlueDatabases(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
+	// TODO put me in helper function
```
Comment (khushboo9024) [inline aws/table_aws_glue_catalog_database.go:89]:
```suggestion
```

File: aws/table_aws_glue_catalog_database.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+	"context"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/service/glue"
+
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+)
+
```
Comment (khushboo9024) [inline aws/table_aws_glue_catalog_database.go:13]:
```suggestion
//// TABLE DEFINITION
```

File: aws-test/tests/aws_glue_catalog_database/test-get-query.sql [hunk 9]
```diff
@@ -0,0 +1,3 @@
+select name, title, akas
```
Comment (khushboo9024) [inline aws-test/tests/aws_glue_catalog_database/test-get-query.sql:1]: can we query other important columns like `catalog_id` , `create_table_default_permissions`, `location_uri`, `description` in get call?

File: aws-test/tests/aws_glue_catalog_database/variables.tf [hunk 10]
```diff
@@ -0,0 +1,70 @@
+
```
Comment (khushboo9024) [inline aws-test/tests/aws_glue_catalog_database/variables.tf:1]: Remove extra line.

File: aws/table_aws_glue_catalog_database.go [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+func tableAwsGlueDatabase(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "aws_glue_catalog_database",
+		Description: "AWS Glue Database",
+		Get: &plugin.GetConfig{
+			KeyColumns:        plugin.SingleColumn("Name"),
+			ShouldIgnoreError: isNotFoundError([]string{"InvalidInputException", "EntityNotFoundException", "InternalServiceException", "OperationTimeoutException", "EncryptionException"}),
```
Comment (khushboo9024) [inline aws/table_aws_glue_catalog_database.go:20]: Do we need all these not found errors ?

File: aws/table_aws_glue_catalog_database.go [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS
+
+func getAwsGlueDatabase(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	logger := plugin.Logger(ctx)
+	logger.Trace("getAwsGlueDatabase")
+
+	// TODO put me in helper function
```
Comment (khushboo9024) [inline aws/table_aws_glue_catalog_database.go:124]:
```suggestion
```

**Review by khushboo9024 — CHANGES_REQUESTED** [review]
Please make suggested changes.

File: aws/table_aws_glue_catalog_database.go [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+	logger := plugin.Logger(ctx)
+	logger.Trace("getAwsGlueDatabase")
+
+	var region string
+	matrixRegion := plugin.GetMatrixItem(ctx)[matrixKeyRegion]
+	if matrixRegion != nil {
+		region = matrixRegion.(string)
+	}
+	var name string
+	name = d.KeyColumnQuals["name"].GetStringValue()
```
Comment (Subhajit97) [inline aws/table_aws_glue_catalog_database.go:130]:
```suggestion
	name := d.KeyColumnQuals["name"].GetStringValue()
```

File: aws/table_aws_glue_catalog_database.go [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+			return !isLast
+		},
+	)
+
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS
+
+func getAwsGlueDatabase(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (Subhajit97) [inline aws/table_aws_glue_catalog_database.go:120]:
```suggestion
func getAwsGlueDatabase(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```

File: aws/table_aws_glue_catalog_database.go [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+	"github.com/aws/aws-sdk-go/service/glue"
+
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+)
+
+//// TABLE DEFINITION
+
+func tableAwsGlueDatabase(_ context.Context) *plugin.Table {
```
Comment (Subhajit97) [inline aws/table_aws_glue_catalog_database.go:15]:
```suggestion
func tableAwsGlueCatalogDatabase(_ context.Context) *plugin.Table {
```

File: aws-test/tests/aws_glue_catalog_database/variables.tf [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+variable "aws_profile" {
+  type    = string
+  default = "default"
+  description = "AWS credentials profile used for the test. Default is to use the default profile."
+}
+
+variable "aws_region" {
+  type    = string
+  default = "us-east-2"
```
Comment (Subhajit97) [inline aws-test/tests/aws_glue_catalog_database/variables.tf:15]: We generally use `us-east-1` as primary region.

File: aws-test/tests/aws_glue_catalog_database/test-hydrate-expected.json [hunk 17]
```diff
@@ -0,0 +1,8 @@
+[
+	{
+        "akas": [
+            "{{ output.resource_aka.value }}"
+        ],
+		"title": "{{ resourceName}}"
```
Comment (Subhajit97) [inline aws-test/tests/aws_glue_catalog_database/test-hydrate-expected.json:6]: Fix indentation

File: docs/tables/aws_glue_catalog_database.md [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+  aws_glue_catalog_database;
+```
+
+
+### Count databases per Catalog ID
+
+```sql
+select
+  catalog_id,
+  count(name)
```
Comment (khushboo9024) [inline docs/tables/aws_glue_catalog_database.md:27]: Please add alias for this column.

File: aws/table_aws_glue_catalog_database.go [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "location_uri",
+				Description: "The location of the database (for example, an HDFS path).",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "create_time",
+				Description: "The time at which the metadata database was created in the catalog.",
```
Comment (khushboo9024) [inline aws/table_aws_glue_catalog_database.go:52]: `create_time` should be third column.

**Review by khushboo9024 — APPROVED** [review]
LGTM

**Review by rajmohanty17 — CHANGES_REQUESTED** [review]
@khushboo9024 can we achieve this in an example ?
Glue Data Catalog Encrypted With KMS Customer Master Keys

**Comment by pskrbasu** [comment]
> @khushboo9024 can we achieve this in an example ?
> Glue Data Catalog Encrypted With KMS Customer Master Keys

@rajmohanty17 `Glue Data Catalog Encrypted With KMS Customer Master Keys` is related to the Data Catalog and not the database. Catalog ID is just the account ID. Databases reside inside Catalog but we can't create any new catalog, only we can encrypt it.
We can add the `AWS KMS Key ID` column in the database table which will have the data if the Catalog(Account ID) is encrypted, but for multiple databases this column will have the same KMS ID.
Is that okay?

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**